File: 500-y.rb
U = ->(f) { f.(f) }
fact = U.(
  ->(f) {
    ->(x) {
      if x == 1
        1
      else
        x * f.(f).(x-1)
      end
    }
  }
)
puts fact.(5)

#Refactor - extract the reapplicaiton of the function to itself inside the factorial function

File: js/500-y.js
const U = (f) => { return f(f) }
const fact = U((f) => {
  return (x) => {
    if (x === 0) {
      return 1
    } else {
      return x * f(f)(x-1)
    }
  }
})
console.log(fact(8))

//Refactor - extract the reapplicaiton of the function to itself inside the factorial function
